packages/notebook/src/widget.ts

```diff
@@ -1350,12 +1369,101 @@ class ScrollbarItem implements WindowedList.IRenderer.IScrollbarItem {
     this._isDisposed = true;
     this._notebook.activeCellChanged.disconnect(this._updateActive);
     this._notebook.selectionChanged.disconnect(this._updateSelection);
+    if (this._model.type === 'code') {
+      const model = this._model as ICodeCellModel;
+      if (model.outputs) {
+        model.outputs.changed.disconnect(this._updateOutput);
+        model.stateChanged.disconnect(this._updateState);
+      }
+    }
   };
 
-  private _createElement() {
-    return document.createElement('li');
+  private _updateState = (
+    _: ICellModel,
+    change: IChangedArgs<
+      any,
+      any,
+      'trusted' | 'isDirty' | 'executionCount' | 'executionState'
+    >
+  ) => {
+    switch (change.name) {
+      case 'executionCount':
+      case 'executionState':
+        this._updatePrompt();
+        break;
+      case 'isDirty': {
+        this._updateDirty();
+        break;
+      }
+    }
+  };
+
+  private _updateDirty() {
+    if (this._model.type !== 'code' || !this._element) {
+      return;
+    }
+    const model = this._model as ICodeCellModel;
+    const wasDirty = this._element.classList.contains(DIRTY_CLASS);
+    if (wasDirty !== model.isDirty) {
+      if (model.isDirty) {
+        this._element.classList.add(DIRTY_CLASS);
+      } else {
+        this._element.classList.remove(DIRTY_CLASS);
+      }
+    }
   }
 
+  private _updateOutput = () => {
+    this._updatePrompt();
+  };
```

Comment:
Can't we use an arrow function for `_updatePrompt()` instead of using this intermediate `_updateOutput` ?

Reply:
Done in [5a77107](https://github.com/jupyterlab/jupyterlab/pull/16432/commits/5a77107e46de8662ed7c94a175f0df7e8d873478)